counsel.el

```diff
@@ -2785,7 +2785,13 @@ A is the left hand side, B the right hand side."
 
 (defun counsel-yank-pop-action-remove (s)
   "Remove S from the kill ring."
-  (setq kill-ring (delete s kill-ring)))
+  (let ((next-ptr (cdr (member s kill-ring))))
+    (setq kill-ring (delete s kill-ring))
+    (setq kill-ring-yank-pointer (or next-ptr kill-ring))))
+
+(defun counsel-yank-pop-action-position (s)
+  "Set yank position to S."
+  (setq kill-ring-yank-pointer (member s kill-ring)))
```

Comment:
In Emacs this is called "rotating" the yank pointer, so I think we should also say "rotate" instead of "position".

Modifying the kill ring and its pointer directly is not usually kosher, as killing and yanking are very central, configurable and flexible features of Emacs. We should, therefore, delegate as much as possible to internal functions:

```el
(defun counsel--yank-pop-position (s)
  "Return position of S in `kill-ring' relative to last yank.
S must exist in `kill-ring'."
  (or (cl-position s kill-ring-yank-pointer :test #'equal-including-properties)
      (+ (- (length kill-ring-yank-pointer)
            (length kill-ring))
         (cl-position s kill-ring :test #'equal-including-properties))))

(defun counsel-yank-pop-action-rotate (s)
  "Rotate the yanking point to S in the kill ring.
See `current-kill' for how this interacts with the window system
selection."
  (current-kill (counsel--yank-pop-position s)))
```

Adding `counsel--yank-pop-position` would make this PR a borderline non-trivial contribution (right @abo-abo ?), so you may need to assign your copyright to the FSF first. Otherwise you can hope that and wait until my next PR lands which needs to define `counsel--yank-pop-position` anyway. :)

Reply:
The promised PR now lives at #1356.

Reply:
I don't have any great desire to do that paperwork, otherwise I'd probably contribute more to this project, in fact I cut this PR down to make it fit under the total limit.  So, anything you can do first I'm OK with.  I'll examine your suggestions later, I'm preoccupied atm.  But you are probably right that it could be simpler.  I'm fairly new to writing [e]lisp code.